src/main/java/org/opensearch/neuralsearch/search/query/HybridQueryPhaseSearcher.java

```diff
@@ -110,29 +115,84 @@ private void setTopDocsInQueryResult(
+            // find any valid doc Id, or set it to -1 if there is not a single match
+            int delimiterDocId = topDocs.stream()
+                .filter(Objects::nonNull)
+                .filter(topDoc -> Objects.nonNull(topDoc.scoreDocs))
+                .map(topDoc -> topDoc.scoreDocs)
+                .filter(scoreDoc -> scoreDoc.length > 0)
+                .map(scoreDoc -> scoreDoc[0].doc)
+                .findFirst()
+                .orElse(-1);
+            if (delimiterDocId == -1) {
+                return new TopDocs(totalHits, scoreDocs);
+            }
+            // format scores using following template:
+            // doc_id | magic_number_1
+            // doc_id | magic_number_2
+            // ...
+            // doc_id | magic_number_2
+            // ...
+            // doc_id | magic_number_2
+            // ...
+            // doc_id | magic_number_1
+            List<ScoreDoc> result = new ArrayList<>();
+            result.add(createStartStopElementForHybridSearchResults(delimiterDocId));
+            for (TopDocs topDoc : topDocs) {
+                if (Objects.isNull(topDoc) || Objects.isNull(topDoc.scoreDocs)) {
+                    result.add(createDelimiterElementForHybridSearchResults(delimiterDocId));
+                    continue;
+                }
+                result.add(createDelimiterElementForHybridSearchResults(delimiterDocId));
+                result.addAll(Arrays.asList(topDoc.scoreDocs));
+            }
+            result.add(createStartStopElementForHybridSearchResults(delimiterDocId));
+            scoreDocs = result.stream().map(doc -> new ScoreDoc(doc.doc, doc.score, doc.shardIndex)).toArray(ScoreDoc[]::new);
+        }
+        return new TopDocs(totalHits, scoreDocs);
+    }
+
+    private TotalHits getTotalHits(final SearchContext searchContext, final List<TopDocs> topDocs, final boolean isSingleShard) {
         int trackTotalHitsUpTo = searchContext.trackTotalHitsUpTo();
         final TotalHits.Relation relation = trackTotalHitsUpTo == SearchContext.TRACK_TOTAL_HITS_DISABLED
             ? TotalHits.Relation.GREATER_THAN_OR_EQUAL_TO
             : TotalHits.Relation.EQUAL_TO;
-        if (topDocs == null || topDocs.size() == 0) {
+        if (topDocs == null || topDocs.isEmpty()) {
             return new TotalHits(0, relation);
         }
         long maxTotalHits = topDocs.get(0).totalHits.value;
+        int totalSize = 0;
         for (TopDocs topDoc : topDocs) {
             maxTotalHits = Math.max(maxTotalHits, topDoc.totalHits.value);
+            if (isSingleShard) {
+                totalSize += topDoc.totalHits.value + 1;
+            }
         }
+        // add 1 qty per each sub-query and + 2 for start and stop delimiters
+        totalSize += 2;
+        if (isSingleShard) {
+            // for single shard we need to update total size as this is how many docs are fetched in Fetch phase
```

Comment:
Could be due to a lack of context, but how do we handle total size for multi shard?
In my understanding, we update totalSize because fetch is already done for single shard.
But for multi shard, still how do we get correct totalSize including start/stop mark and delimiter count?

Reply:
for multi shards we use total hits from the sub-query with max results. In the processor query results will be updated, and fetch phase will get already well formed results where total hits will match number of ScoreDocs in the list. It's only different for 1 shard case